python/ray/llm/_internal/batch/stages/sglang_engine_stage.py

```diff
@@ -0,0 +1,367 @@
+        **kwargs,
+    ):
+        self.request_id = 0
+        self.idx_in_batch_column = idx_in_batch_column
+        self.task_type = kwargs.pop("task", SGLangTaskType.GENERATE)
+        self.model_path = kwargs.get("model_path", None)
+        assert self.model_path is not None
+
+        if sgl is None:
+            raise ImportError(
+                "SGLang is not installed or failed to import. Please run "
+                "`pip install ray[llm]` to install required dependencies."
+            )
+
+        # Initialize the SGLang engine
+        self.engine = sgl.Engine(**kwargs)
+
+        # The performance gets really bad if there are too many requests in the pending queue.
+        # We work around it with semaphore to limit the number of concurrent requests in the engine.
+        self.max_pending_requests = max_pending_requests
+        if self.max_pending_requests > 0:
+            self.semaphore = asyncio.Semaphore(self.max_pending_requests)
+        else:
+            self.semaphore = asyncio.NullContext()
+
+    async def _prepare_llm_request(self, row: Dict[str, Any]) -> SGLangEngineRequest:
+        """Prepare the inputs for LLM inference.
+
+        Args:
+            row: The row.
+
+        Returns:
+            A single SGLangEngineRequest.
+        """
+        text = row.pop("text")
+        if "input_ids" in row:
+            input_ids = row.pop("input_ids").tolist()
+        else:
+            input_ids = None
+
+        # Prepare sampling parameters.
+        if self.task_type == SGLangTaskType.GENERATE:
+            sampling_params = row.pop("sampling_params")
+        else:
+            raise ValueError(f"Unsupported task type: {self.task_type}")
+
+        request = SGLangEngineRequest(
+            request_id=self.request_id,
+            idx_in_batch=row[self.idx_in_batch_column],
+            text=text,
+            input_ids=input_ids,
+            sampling_params=sampling_params,
+        )
+        self.request_id += 1
+        return request
+
+    async def generate_async(
+        self, row: Dict[str, Any]
+    ) -> Tuple[SGLangEngineRequest, Dict[str, Any]]:
+        """Process a single request.
+
+        Args:
+            request: The request.
+
+        Returns:
+            A tuple of index in batch, request output and bypassed custom fields.
+        """
+        request = await self._prepare_llm_request(row)
+
+        async with self.semaphore:
+            output = await self._generate_async(request)
+
+        output_data = SGLangOutputData.from_sglang_engine_output(output)
+        return request, output_data.model_dump()
+
+    async def _generate_async(self, request: SGLangEngineRequest) -> Any:
+        """Process a single request.
+
+        Args:
+            request: The request.
+
+        Returns:
+            The output of the request.
+        """
+
+        # Send the request to the LLM engine.
+        output = await self.engine.async_generate(
```

Comment:
how does dynamic batching work in this case? in the vllm case we return a stream and return the output as soon as the stream hits a finished state.

Reply:
Thanks for pointing out this! Now I have changed this to a streaming-style API.